### ISSUE TITLE
Date Picker

### ISSUE THREAD

**Issue by SylvainCorlay** [ipython/ipywidgets#168]
The native color picker that is one of the tutorial examples could become a first-class member of ipywidgets, using the Color trait type. Same for a Date picker.

The datetime trait type would probably be slightly more complicated to handle different time zones between browser and kernel...

**Comment by SylvainCorlay** [comment]
For other libraries using the Color trait type like bqplot, it would make it very easy to bind some plot item color with a native ipywidgets color picker.

**Comment by ellisonbg** [comment]
+1 on having Color and Date widgets. I have already needed these - would be
helpful to have date/time widgets for different granulatiries (full time,
day of month, week, day of week, month).

On Tue, Oct 6, 2015 at 7:02 PM, Sylvain Corlay <notifications@github.com>
wrote:

> For other libraries using the Color trait type like bqplot, it would make
> it very easy to bind some plot item color with a native ipywidgets color
> picker.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/issues/168#issuecomment-146028771>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by SylvainCorlay** [comment]
The color picker is the easy one. The date-time trait is more tricky because of time zones in my experience.

**Comment by ellisonbg** [comment]
Yes, for full blown datetime, you have to worry about that - but for the
simpler ones, it isn't much of a problem.

On Tue, Oct 6, 2015 at 9:33 PM, Sylvain Corlay <notifications@github.com>
wrote:

> The color picker is the easy one. The date-time trait is more tricky
> because of time zones in my experience.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/issues/168#issuecomment-146052754>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by SylvainCorlay** [comment]
Widget_color added in #169 .

**Comment by jdfreder** [comment]
@SylvainCorlay why not just sync UTC in the model and assume the clocks are synchronized?  I would not worry about clock synchronization at this level.  The view could render in local timezone instead of UTC.

**Comment by SylvainCorlay** [comment]
I am not quite sure because even a bare date gets conerted to datetime and
then shifted by a few hours.

**Comment by SylvainCorlay** [comment]
I discussed this in person with @pganssle earlier today and since JavaScriptś Date is not timezone-naive, unlike python's datetime, we might have to be opinionated  on de-serialization by adding the local time zone to any datetime coming from the backend that does not have one...

**Comment by pganssle** [comment]
Yeah, I was holding off on commenting until I understood the JS a bit more, it seems like the main issue is that Javascript handles dates a bit differently than python. JS doesn't seem to have any timezone-awareness, all times are assumed to be in local time, according to [this link](http://stackoverflow.com/a/2771647). You apparently also need to use the `new` operator when creating dates, otherwise you'll get a date string (see [the note on this page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)).

Instantiating the Date object with a string seems to assume that the string is a UTC string, while instantiating it with constructor arguments assumes that it's in local time:
```javascript
>>> cdate = new Date('2015-05-12')
Mon May 11 2015 20:00:00 GMT-0400 (Eastern Daylight Time)
>>> cdate = new Date(2015, 4, 12) // Months are 0-based indices in JS
Tue May 12 2015 00:00:00 GMT-0400 (Eastern Daylight Time)
```

It seems to me that the right behavior is that python timezone-naive `datetime`s can be passed directly to Javascript with the standard constructor (resulting in the correct local time), and timezone-aware objects should be converted to UTC in python and passed to the `Date.UTC` function in Javascript:

```javascript
>>> cdate = new Date(Date.UTC(2015, 4, 12)) // Test
Mon May 11 2015 20:00:00 GMT-0400 (Eastern Daylight Time)
```
Here's a demo python script that I think has the right behavior:

```python
def to_JS_datestring(pydt):
    argstr = "{year}, {month}, {day}, {hour}, {minute}, {second}, {millisecond}"
    if pydt.tzinfo is not None:
        # Naive python object
        pydt = (pydt - pydt.utcoffset()).replace(tzinfo=None)
        argstr = "Date.UTC({args})".format(args=argstr)

    args = argstr.format(year=pydt.year, month=pydt.month - 1,
                         day=pydt.day, hour=pydt.hour,
                         minute=pydt.minute, second=pydt.second,
                         millisecond=pydt.microsecond * 1000)

    return "new Date({argstr})".format(argstr=args)

def test_JS(ds, vname):
    o = []
    for ii, d in enumerate(ds):
        o.append("{vn}{x} = {s}".format(vn=vname,
                                        x=ii,
                                        s=to_JS_datestring(d)))

    o.append(vname + " = [" +
             ', '.join("{vn}{x}".format(vn=vname, x=x)
                       for x in range(len(ds))) + "];")

    o.append("for (i = 0; i < " + vname + ".length - 1; i++) {")
    o.append("    console.log(" + vname + "[i] - " + vname + "[i+1] == 0);")
    o.append("}")

    return '\n'.join(o)


if __name__ == "__main__":
    from datetime import datetime as dt
    from dateutil.tz import gettz, tzutc

    utc = tzutc()
    local_tz = gettz('America/New_York')
    minsk_tz = gettz('Europe/Minsk')

    dbase = dt(2015, 4, 13)
    dtbase = dt(2015, 4, 13, 3, 15, 21, 150000)

    # These should all be the same value in the output
    dlocal = dbase.replace(tzinfo=local_tz)
    dutc = dlocal.astimezone(utc)
    dminsk = dlocal.astimezone(minsk_tz)

    ds = (dbase, dlocal, dutc, dminsk)

    print(test_JS(ds, "dates"))

    # These should also all be the same value in the output
    dtlocal = dtbase.replace(tzinfo=local_tz)
    dtutc = dtlocal.astimezone(utc)
    dtminsk = dtlocal.astimezone(minsk_tz)

    dts = (dtbase, dtlocal, dtutc, dtminsk)

    print("\n" + test_JS(dts, "datetimes"))
```

This outputs:
```javascript
dates0 = new Date(2015, 3, 13, 0, 0, 0, 0)                                   
dates1 = new Date(Date.UTC(2015, 3, 13, 4, 0, 0, 0))                         
dates2 = new Date(Date.UTC(2015, 3, 13, 4, 0, 0, 0))                         
dates3 = new Date(Date.UTC(2015, 3, 13, 4, 0, 0, 0))                         
dates = [dates0, dates1, dates2, dates3];                                    
for (i = 0; i < dates.length - 1; i++) {                                     
    console.log(dates[i] - dates[i+1] == 0);                                 
}                                                                            
                                                                             
datetimes0 = new Date(2015, 3, 13, 3, 15, 21, 150000000)                     
datetimes1 = new Date(Date.UTC(2015, 3, 13, 7, 15, 21, 150000000))           
datetimes2 = new Date(Date.UTC(2015, 3, 13, 7, 15, 21, 150000000))           
datetimes3 = new Date(Date.UTC(2015, 3, 13, 7, 15, 21, 150000000))           
datetimes = [datetimes0, datetimes1, datetimes2, datetimes3];                
for (i = 0; i < datetimes.length - 1; i++) {                                 
    console.log(datetimes[i] - datetimes[i+1] == 0);                         
}                                                                            
```

Hope that's helpful.

**Comment by SylvainCorlay** [comment]
Thanks again @pganssle I added a date picker widgets, and your comments helped a lot. This will be part of soon-to-be-released ipywidgets 6.0
